### PR TITLE
Release instructions

### PR DESCRIPTION
**Pull request by altendky**

### CODE REVIEW

File: CONTRIBUTING.rst
Comment (altendky) [inline CONTRIBUTING.rst:54]: I haven't used this much so I don't know any nuances.  Should this just be used for all tagging of release candidates and final releases?  Will we have to manually add the sdist and wheel artifacts?

File: CONTRIBUTING.rst
Comment (altendky) [inline CONTRIBUTING.rst:60]: The Twisted maillist that is...  I dunno.

File: CONTRIBUTING.rst
Comment (altendky) [inline CONTRIBUTING.rst:76]: Probably ought to just leave this for a separate activity I guess.  It isn't needed to do a release and isn't presently being done at all.

File: CONTRIBUTING.rst
Comment (adiroiban) [inline CONTRIBUTING.rst:8]:
```suggestion
- Create a commit for a release candidate inside the release branch (see next for details
- Create a Pull Request for the release. The same branch and PR will be used for both the release candidate and  the final release.
- Get an approving review
- Once approved, tag the last commit in the branch as a release candidate.
- Push the tag and this trigger the upload of artifacts to PyPI.
```

File: CONTRIBUTING.rst
Comment (adiroiban) [inline CONTRIBUTING.rst:9]: Where to notify? I have no idea where to notify. Maybe IRC #twisted-dev  ?

How much time? -- Maybe 1 week.

File: CONTRIBUTING.rst
Comment (adiroiban) [inline CONTRIBUTING.rst:10]: My take on this.

```suggestion
- Address concerns raised via new issues and fix them in dedicated branches that will be first merged in the main branch.
- Once merged in the main branch, cheery-pick the merge commit in the release branch.
```

File: CONTRIBUTING.rst
Comment (adiroiban) [inline CONTRIBUTING.rst:11]:
```suggestion
- Prepare a commit for a final release and push it to the release branch
- Get an approving review for the final release.
- Tag the commit using the final release version. This will trigger the upload of artifacts to PyPI.
```

File: CONTRIBUTING.rst
Comment (adiroiban) [inline CONTRIBUTING.rst:34]:
```suggestion
  - On the new release branch you will commit all tagged release candidate commits as well as the final tagged release commit.
```

File: CONTRIBUTING.rst
Comment (adiroiban) [inline CONTRIBUTING.rst:40]:
```suggestion
- Run ``venv/bin/towncrier build --yes`` to build the the newsfragments into the release notes document and to automatically remove the newsfragment files.
```

File: CONTRIBUTING.rst
Comment (adiroiban) [inline CONTRIBUTING.rst:44]: I think that we can keep it simple. One reason is for tags, the other reason is that forks don't have access to PYPI secrets.


```suggestion
- Commit and push to the primary repository, not a fork.
```

File: CONTRIBUTING.rst
Comment (adiroiban) [reply CONTRIBUTING.rst:54]: I prefer using git CLI to tag.
For small projects I don't think that we need to GItHub release GUI.

File: CONTRIBUTING.rst
Comment (adiroiban) [reply CONTRIBUTING.rst:60]: mailing list can work and IRC for quick feedback...it's very low trafic

File: CONTRIBUTING.rst
Comment (adiroiban) [inline CONTRIBUTING.rst:64]: my idea is to fix master first...and then the release branch... but this is also ok.

**Review by adiroiban — COMMENTED** [review]
just initial comments

File: CONTRIBUTING.rst
Comment (adiroiban) [reply CONTRIBUTING.rst:76]: Good question. I think that we should have a placeholder like. "No changes since last release."

For my project, I do a stupid think and replace the release candidate in the release notes with the final note.
But this erases the project history and I think it's bad practice.

In my fork of towncrier, If I run towncrier with a new version and no news fragments I get an em

But we can have something like this for the default template

```
{% set underline = "^" %}
{% set section = sections[''] %}
{% set empty_release = True %}

{% for category, val in definitions.iteritems()%}
{% if definitions[category]['showcontent'] and section[category]|length > 0 %}
{{ definitions[category]['name'] }}
{{ underline * definitions[category]['name']|length }}

{% for text, values in section[category]|dictsort(by='value') %}
* {{ text }} {{ values|sort|join(', ') }}
{% set empty_release = False %}
{% endfor %}


{% endif %}
{% endfor %}
{% if empty_release %}
No changes since the last release.
{% endif %}
```

To produce something like

```

Version 4.8.0, 2021-02-01
-------------------------


No changes since the last release.


Version 4.8.0rc1, 2020-11-19
-------------------------


New Features
^^^^^^^^^^^^

* The embedded OpenSSL libraries used on Windows, macOS, and generic Linux were
  updated to version 1.1.1h. [#5496]
```

**Review by adiroiban — CHANGES_REQUESTED** [review]
Looks like a very good start. Thanks. I left my comments.
You are the boss so feel free to do it as you like :)

File: CONTRIBUTING.rst
Comment (altendky) [reply CONTRIBUTING.rst:60]:
```suggestion
- Notify the `Twisted-Python maillist <https://twistedmatrix.com/cgi-bin/mailman/listinfo/twisted-python>`_ of the release to allow for feedback if a pre-release and just for notification if a final release.
```

File: CONTRIBUTING.rst
Comment (altendky) [reply CONTRIBUTING.rst:64]: I guess as long as the cherry pick from master is done as a PR against the release branch it doesn't really matter.  My point is to encourage that the actual release branch stays reasonably clean to alleviate confused tossing of changes there and flailing around trying to fix something.

```suggestion
  - Submit PRs against the release branch to integrate the needed changes.  Any PRs could be cherry picks from the ``master`` branch if already resolved there, or direct PRs against the release branch that will be merged back into master at the completion of the release.
```

File: CONTRIBUTING.rst
Comment (altendky) [reply CONTRIBUTING.rst:44]: Explaining why is detrimental here?  hmm...

```suggestion
- Commit and push to the primary repository, not a fork.

  - It is important to not use a fork so that pushed tags end up in the primary repository, server provided secrets for publishing to PyPI are available, and maybe more.
```

File: CONTRIBUTING.rst
Comment (altendky) [reply CONTRIBUTING.rst:8]: I Think the summary section is sufficiently short and well labeled to not need to reference ahead for details in the middle.  Also, other small adjustments.

```suggestion
- Create a commit for a release candidate inside the release branch.
- Create a Pull Request for the release. The same branch and PR will be used for both the release candidates and  the final release.
- Get an approving review.
- Once approved, tag the last commit in the branch as a release candidate.
- Push the tag.  This will trigger a build including the upload of artifacts to PyPI.
```

File: CONTRIBUTING.rst
Comment (altendky) [reply CONTRIBUTING.rst:9]:
```suggestion
- Notify the Twisted maillist allow a week for feedback before continuing.
```

File: CONTRIBUTING.rst
Comment (altendky) [reply CONTRIBUTING.rst:10]:
```suggestion
- Make sure that tickets exist for all raised concerns.  These tickets should be marked using the ``blocking`` label.
- Address each issue by either concluding that it is not really a release blocker and removing the label or by fixing via a PR to the release branch.
```

File: CONTRIBUTING.rst
Comment (altendky) [reply CONTRIBUTING.rst:11]:
```suggestion
- Prepare a commit for a final release and push it to the release branch.
- Get an approving review for the final release.
- Tag the commit using the final release version. This will trigger the upload of artifacts to PyPI.
```

File: CONTRIBUTING.rst
Comment (altendky) [reply CONTRIBUTING.rst:76]: I was referring to the project version, unrelated to generating release notes.  Rather, if I checkout and install an arbitrary commit of towncrier I will get one claiming to be the last release version since after doing a release the version is just left alone.

Despite having numerous changes since the release of 19.2, the latest commit on `master` still claims to be 19.2.

https://github.com/twisted/towncrier/blame/00f19b4d24ad5dfd57124f7c93f9ba2107408f8c/src/towncrier/_version.py#L10

**Comment by adiroiban** [comment]
> To keep master up to date, perhaps the release branch should be merged back into master not only at the end but also after each pre-release?

I would say it's not worth the trouble.
But it's your call :)


> I was referring to the project version, unrelated to generating release notes. Rather, if I checkout and install an arbitrary commit of towncrier I will get one claiming to be the last release version since after doing a release the version is just left alone.

Sorry for missing that.

Yes. I think that after the public release there should another step to update the release to +1 and + .dev0

so if 21.1.0 was release, after the release the release branch is updated with 21.1.1dev0 and merged into master.

For a 2.1.2 release... it will be updated to 2.1.3dev0

**Comment by altendky** [comment]
this is getting painful...  not complaining about your reviews, just the detail and growth of the process and a minimum of triplicate reviews and dismissing reviews and...  the review dismissals make me wonder if it would be simpler to treat each sub-release via a separate PR against the release branch.  sounds complicated, also sounds simpler since you wouldn't dismiss reviews.  anyways.  :|

File: CONTRIBUTING.rst
Comment (adiroiban) [inline CONTRIBUTING.rst:80]: Maybe keep it simple.

I don't think that we need to go into the `edit src/towncrier/_version.py` as by this point people should know how do update the version.

```suggestion
- If the final release has been completed, mark current version as development `Version('towncrier', 19, 9, 1, dev=0)`` and merge it without requesting another review.
```

File: CONTRIBUTING.rst
Comment (altendky) [inline CONTRIBUTING.rst:88]:
```suggestion
- Merge without waiting for an approving review.
```

**Review by adiroiban — APPROVED** [review]
All good. Thanks!

I think that having a single PR is easier. I guess that we can give it a try and see if things get ugly.. .then the process can be changed with a PR for each release and separate PR for each release candidates.

I think the title of  this PR can be updated and this PR merged .

Looking forward to the next release where we can practice what is documented here.

Many thanks again for putting this together.

**Comment by altendky** [comment]
Thanks for seeing it through with the reviews.

**Pull request merged**
